pycbc/distributions/constraints.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (C) 2017 Christopher M. Biwer
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation; either version 3 of the License, or (at your
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+"""
+This modules provides classes for evaluating multi-dimensional constraints.
+"""
+
+from pycbc import transforms
+
+class Constraint(object):
+    """ Creates a constraint that evaluates to True if parameters obey
+    the constraint and False if they do not.
+    """
+    name = "custom"
+    required_parameters = []
+    def __init__(self, variable_args, **kwargs):
+
+        # set any given attributes and get transforms from variable_args
+        # to required parameters
+        for kwarg in kwargs.keys():
+            setattr(self, kwarg, kwargs[kwarg])
+        _, self.transforms = transforms.get_common_cbc_transforms(
+                                       self.required_parameters, variable_args)
+
+    def __call__(self, params):
+        """ Evaluates constraint.
+        """
+        params = transforms.apply_conversions(params, self.transforms) \
+                     if self.transforms else params
+        return self._constraint(params)
+
+    def _constraint(self, params):
+        """ Evaluates constraint function.
+        """
+        return eval(self.func)
```

Comment:
I don't like having an eval here. We should limit how much these are used. I think it'd be better to use the eval that's in the guts of `FieldArray`. So, in `__call__`, before `apply_transforms`:
```
if isinstance(params, dict):
   params = FieldArray.from_kwargs(**params)
elif not isinstance(params, FieldArray):
   raise ValueError("params must be dict or FieldArray instance")
```
Then after the transforms are applied:
```
out = self._constraint(params)
if out.size == 1:
    out = out.item()
return out
```
Finally, change `_constraint` to:
```
return params[self.func]
```
This means that in the config file, you would only have to put:
```
[constraint-mchirp]
func = (mchirp >= 9.5) & (mchirp < 10.5)
```
To constrain the parameter in some bounds.

Although rename the `func` attribute to be more descriptive name. Maybe `constraint_arg`?